src/pyats/contrib/plugins/topoup_plugin/topoup.py

```diff
@@ -0,0 +1,139 @@
+                         action='store',
+                         default=120,
+                         help='Total time allowed for checking devices connectivity')
+
+        # -connection_check_interval
+        # --connection_check_interval
+        grp.add_argument(*connection_check_interval,
+                         dest='connection_check_interval',
+                         action='store',
+                         default=10,
+                         help='Time to sleep between device connectivity checks')
+
+        return grp
+
+
+    def pre_job(self, task):
+        '''Loop over all the topology devices and make sure they are up and running
+           before executing the test script.
+        '''
+
+        # Set the timers
+        start_time = time()
+        timeout = self.runtime.args.connection_check_timeout
+        interval = self.runtime.args.connection_check_interval
+
+        log.info("Connectivity check timeout is '{timeout}' and "
+            "connectivity check interval is '{interval}'".format(timeout=timeout, interval=interval))
+
+        # Looping over devices and make sure they are up
+        pcall_output = pcall(device_connect,
+            ckwargs = {'start_time': start_time, 'timeout': timeout, 'interval': interval},
+            ikwargs = [{'device':self.runtime.testbed.devices[dev]} for dev in self.runtime.testbed.devices])
+
+        if not (pcall_output[0] and pcall_output[1]):
+            # Terminate testscript
+            raise Exception ("Not all the testbed devices are up and ready")
+
+        return
+
+
+def device_connect(device, start_time, timeout, interval):
+    '''Try to connect to the device and if fails, sleep for interval seconds and retry
+       till the timeout is reached
+
+        Args:
+            device ('obj'): device to use
+            start_time ('int'): Current time to calculate the timeout, seconds
+            timeout ('int'): Timeout value when reached exit even if failed, seconds
+            interval ('int'): Sleep time between retries, seconds
+
+        Returns:
+            result(`bool`): Device is successfully connected
+
+        Raises:
+            None
+
+    '''
+
+    time_difference = time() - start_time
+
+    while (time() - start_time) < float(timeout):
+
+        try:
+            # Connect to the device
+            device.connect()
+            log.info("Successfully connected to '{device}'".format(device=device))
+
+            # Return the pcall call with True
+            return True
+
+        except:
+            # Not ready sleep and retry
+            log.info("Sleeping for '{interavl}' seconds and retry, remaining time {remaining_time}".format(
+                interavl=interval, remaining_time=timeout-time_difference))
+
+            # Sleep for `interval` seconds
+            sleep(interval)
+
+            # Retry connecting to the device
+            device_connect(device, start_time, timeout, interval)
```

Comment:
Why call yourself again; but without return.

If it fails first time, it goes here, it pass, but didnt return, go back to top of while; try again, etc. You will connect multiple times.

Just make it a while until max time run out, and if run out return False else return True

Reply:
Yeah that was the recursive first design before using pcall, make sense for sure. Thanks!

Reply:
```
    def pre_job(self, task):
        '''Loop over all the topology devices and make sure they are up and running
           before executing the test script.
        '''

        # Looping over devices and make sure they are up
        start_time = time()
        timeout = self.runtime.args.connection_check_timeout
        interval = self.runtime.args.connection_check_interval
        log.info("Connectivity check timeout is '{timeout}' and "
            "connectivity check interval is '{interval}'".format(timeout=timeout, interval=interval))
        devices_connect(self.runtime.testbed.devices, start_time, timeout, interval)


def devices_connect(device_dict, start_time, timeout, interval):
    '''
    '''

    while (time() - start_time) < float(timeout):
        for device in device_dict:
            try:
                device_dict[device].connect()
            except:
                # Not ready sleep and retry
                log.info("Sleeping for '{interavl}' seconds and retry".format(interavl=interval))
                sleep(interval)
                devices_connect(device_dict, start_time, timeout, interval)
            else:
                log.info("Successfully connected to '{device}'".format(device=device))
                devices_connect({i:device_dict[i] for i in device_dict if i!=device},
                    start_time, timeout, interval)

        return True

    return False
```